source/pervasive/nonlinear_arith/internals/div_internals.rs

```diff
@@ -0,0 +1,233 @@
+        assert((i + (j + n)) / n == ((i + j) + n) / n);
+        assert(((i - n) + j) / n == ((i + j) - n) / n);
+        assert((i + (j - n)) / n == ((i + j) - n) / n);
+    }
+
+    lemma_mod_induction_forall2(n, f);
+    assert(f(x, y));
+    }
+
+    assert forall |x:int, y:int|
+        {let z = (x % n) - (y % n);
+        (  (0 <= z < n && #[trigger]((x - y) / n) == x / n - y / n)
+            || (-n <= z < 0  && #[trigger]((x - y) / n) == x / n - y / n - 1))} by
+    {
+    let f = |xx:int, yy:int|
+        {let z = (xx % n) - (yy % n);
+            (  (0 <= z < n &&((xx - yy) / n) == xx / n - yy / n)
+                || (-n <= z < 0 && (xx - yy) / n == xx / n - yy / n - 1))};
+    
+    assert forall |i: int, j: int| {
+        &&& ( j >= 0 && #[trigger]f(i, j) ==> f(i, j + n))
+        &&& ( i < n  && f(i, j) ==> f(i - n, j))
+        &&& ( j < n  && f(i, j) ==> f(i, j - n))
+        &&& ( i >= 0 && f(i, j) ==> f(i + n, j))
+    } by
+    {
+        assert(((i + n) - j) / n == ((i - j) + n) / n);
+        assert((i - (j - n)) / n == ((i - j) + n) / n);
+        assert(((i - n) - j) / n == ((i - j) - n) / n);
+        assert((i - (j + n)) / n == ((i - j) - n) / n);
+    }
+    assert forall |i: int, j: int| 0 <= i < n && 0 <= j < n ==> #[trigger]f(i, j) by
+    {
+        assert(((i + n) - j) / n == ((i - j) + n) / n);
+        assert((i - (j - n)) / n == ((i - j) + n) / n);
+        assert(((i - n) - j) / n == ((i - j) - n) / n);
+        assert((i - (j + n)) / n == ((i - j) - n) / n);
+    }
+    lemma_mod_induction_forall2(n, f);
+    assert(f(x, y));
+    }
+}
+
+spec fn add(x: int, y: int) -> int
+{
+    x + y
+}
+
+spec fn sub(x: int, y: int) -> int
+{
+    x - y
+}
+
+/// Performs auto induction for division 
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_induction_auto(n: int, x: int, f: FnSpec(int) -> bool)
+    requires
+        n > 0,
+        div_auto(n) ==>{&&& (forall |i: int| #[trigger]is_le(0, i) && i < n ==> f(i))
+                        &&& (forall |i: int| #[trigger]is_le(0, i) && f(i) ==> f(i + n))
+                        &&& (forall |i: int| #[trigger]is_le(i + 1, n) && f(i) ==> f(i - n))}
+    ensures  
+        div_auto(n),
+        f(x)
+{
+    lemma_div_auto(n);
+    assert(forall |i: int| is_le(0, i) && i < n ==> f(i));
+    assert(forall |i: int| is_le(0, i) && #[trigger]f(i) ==> #[trigger]f(add(i, n)));
+    assert(forall |i: int| is_le(i + 1, n) && #[trigger]f(i) ==> #[trigger]f(sub(i, n)));
+    assert forall |i: int| 0 <= i < n implies #[trigger]f(i) by {
+        assert(f(i)) by {
+            assert(forall |i: int| is_le(0, i) && i < n ==> f(i));
+            assert(is_le(0, i) && i < n);
+        };
+    };
+    assert forall |i: int| i >= 0 && #[trigger]f(i) ==> #[trigger]f(crate::nonlinear_arith::internals::mod_internals::add(i, n)) by {
+        assert(crate::nonlinear_arith::internals::mod_internals::add(i, n) == add(i, n));
+    }; // OBSERVE COMMUNICATION
```

Comment:
Out of curiosity, what does "OBSERVE COMMUNICATION" mean? I'm aware of "OBSERVE" comments, but not sure of the "COMMUNICATION" bit